olive/engine/engine.py

```diff
@@ -393,21 +401,33 @@ def run_no_search(
             signal,
             model_ids,
         ) = self._run_passes(next_step["passes"], model, model_id, accelerator_spec)
-        model_id = model_ids[-1]
-
-        prefix_output_name = f"{output_name}_{accelerator_spec}_" if output_name is not None else f"{accelerator_spec}_"
-        # save the model to output_dir
-        output_model_name = f"{prefix_output_name}model"
-        output_model_json = cache_utils.save_model(
-            model_number=model_id,
-            output_dir=output_dir,
-            output_name=output_model_name,
-            overwrite=True,
-            cache_dir=self._config.cache_dir,
-        )
+        # names of the output models of the passes
+        pass_output_names = [self.passes[pass_name]["output_name"] for pass_name, _ in next_step["passes"]]
+        pass_output_names = [f"{name}_{accelerator_spec}" if name else None for name in pass_output_names]
+
+        final_output_name = pass_output_names[-1]
+        if output_name:
+            # override the output name of the last pass
+            logger.debug("Engine output_name is provided. Will ignore output_name for final pass")
+            final_output_name = f"{output_name}_{accelerator_spec}"
+        elif not final_output_name:
+            # use the default output name
+            final_output_name = str(accelerator_spec)
+        pass_output_names[-1] = final_output_name
+
+        for pass_output_name, pass_output_model_id in zip(pass_output_names, model_ids):
+            if not pass_output_name:
+                continue
+            output_model_json = cache_utils.save_model(
+                model_number=pass_output_model_id,
+                output_dir=output_dir,
+                output_name=f"{pass_output_name}_model",
+                overwrite=True,
+                cache_dir=self._config.cache_dir,
```

Comment:
need these model be recorded into footprints? if so, we could merge with search.

Reply:
There is no need to use footprints here. No-search mode doesn't use footprints.